msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -554,51 +557,127 @@ public void run() {
+                    TAG + methodName,
+                    "Eligible to use broker. Try to get the account from local and broker cache."
+            );
+            getBrokerAndLocalAccount(oid, callback);
         } else {
-            com.microsoft.identity.common.internal.logging.Logger.warn(
+            // load local account and find the matching account
+            com.microsoft.identity.common.internal.logging.Logger.info(
                     TAG + methodName,
-                    "Provided authority was not AAD - defaulting to parsed home_account_id"
+                    "Not eligible to use broker. Try to get the account from local cache."
             );
+            getLocalAccount(oid, callback);
         }
+    }
 
-        AccountRecord accountToReturn = null;
+    private void getBrokerAndLocalAccount(@NonNull final String oid,
+                                          @NonNull final AccountsLoadedCallback<IAccount> callback) {
+        final String methodName = ":getBrokerAndLocalAccount";
 
-        if (null != realm) {
-            accountToReturn = AccountAdapter.getAccountInternal(
-                    mPublicClientConfiguration.getClientId(),
-                    mPublicClientConfiguration.getOAuth2TokenCache(),
-                    homeAccountIdentifier,
-                    realm
-            );
-        } else {
-            com.microsoft.identity.common.internal.logging.Logger.warn(
+        final Handler handler = getPreferredLooper();
+
+        //Make a background call to get account from broker
+        new BrokerMsalController().getBrokerAccounts(
+                mPublicClientConfiguration,
+                new BrokerAccountsLoadedCallback() {
+                    @Override
+                    public void onAccountsLoaded(final List<AccountRecord> accountRecords) {
+                        final List<AccountRecord> cachedAccounts = new ArrayList<>();
+                        cachedAccounts.addAll(getLocalAccounts());
+                        cachedAccounts.addAll(accountRecords);
+
+                        if (cachedAccounts.isEmpty()) {
+                            com.microsoft.identity.common.internal.logging.Logger.info(
+                                    TAG + methodName,
+                                    "No account stored in the broker cache."
+                            );
+                            handler.post(new Runnable() {
+                                @Override
+                                public void run() {
+                                    //Return null if no account found.
+                                    callback.onAccountsLoaded(null);
+                                }
+                            });
+
+                            return;
+                        }
+
+                        for (final AccountRecord accountRecord : cachedAccounts) {
+                            if (!StringUtil.isEmpty(accountRecord.getLocalAccountId())
```

Comment:
This loop and its conditions are duplicated across methods [here ](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/8709d0d6214f9e67a778096fcf102801752bc8ff/msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java#L626-L643)and [here](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/8709d0d6214f9e67a778096fcf102801752bc8ff/msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java#L663-L673) -- we should split this into a few different functions that we can call again and again, rather than duplicate --[ I have specific suggestions here](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/592/files#r281266640).